File: 백준/Silver/24480. 알고리즘 수업 － 깊이 우선 탐색 2/알고리즘 수업 － 깊이 우선 탐색 2.cc
#include <iostream>
#include <algorithm>
#include <vector>

using namespace std;

int N, M, R;
int cnt = 1;
vector<vector<int>> adj;
vector<int> visited;

void dfs(int start) {
    visited[start] = cnt;
    for (auto num : adj[start]) {
        if (visited[num] == 0) {
            cnt++;
            dfs(num);  // 세미콜론 추가!
        }
    }
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(NULL);

    cin >> N >> M >> R;

    adj.resize(N + 1);
    visited.resize(N + 1, 0);

    for (int i = 0; i < M; i++) {
        int a, b;
        cin >> a >> b;
        adj[a].push_back(b);
        adj[b].push_back(a);
    }

    // 내림차순 정렬
    for (int i = 1; i <= N; i++) {
        sort(adj[i].begin(), adj[i].end(), greater<int>());
    }

    dfs(R);

    for (int i = 1; i <= N; i++) {
        cout << visited[i] << '\n';
    }

    return 0;
}

File: 백준/Silver/24480. 알고리즘 수업 － 깊이 우선 탐색 2/알고리즘 수업 － 깊이 우선 탐색 2.py
import sys
sys.setrecursionlimit(10**6)
input = sys.stdin.readline  # ✅ 빠른 입력 처리

N, M, R = map(int, input().split())

adj = [[] for _ in range(N+1)]

for _ in range(M):
    a, b = map(int, input().split())
    adj[a].append(b)
    adj[b].append(a)

for i in range(1, N+1):
    adj[i].sort(reverse=True)  # 내림차순으로 정점 방문

visited = [0] * (N+1)
cnt = 1

def dfs(start):
    global cnt
    visited[start] = cnt
    for num in adj[start]:
        if not visited[num]:
            cnt += 1
            dfs(num)

dfs(R)

for i in range(1, N+1):
    print(visited[i])
